frontend/exporter/src/types/copied.rs

```diff
@@ -3269,16 +3284,29 @@ pub struct Clause {
     pub id: u64,
 }
 
-impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, Clause> for rustc_middle::ty::Clause<'tcx> {
+impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, Clause>
+    for rustc_middle::ty::Binder<'tcx, rustc_middle::ty::ClauseKind<'tcx>>
+{
     fn sinto(&self, s: &S) -> Clause {
-        use rustc_middle::ty::ToPredicate;
+        // FIXME: `ClauseKind::sinto` uses a dummy binder. We need it because `Predicate::sinto()`
+        // doesn't propagate the binder to the `ClauseKind`. This might cause inconsistencies. It
+        // might be that we don't want to hash the binder at all.
+        let binder: Binder<ClauseKind> = self.sinto(s);
+        let id = clause_id_of_bound_clause_kind(&binder);
         Clause {
-            id: clause_id_of_predicate(s, self.clone().to_predicate(s.base().tcx)),
-            kind: self.sinto(s),
+            kind: binder.value,
+            id,
         }
     }
 }
 
+impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, Clause> for rustc_middle::ty::ClauseKind<'tcx> {
+    fn sinto(&self, s: &S) -> Clause {
+        // FIXME: this is dangerous.
+        rustc_middle::ty::Binder::dummy(self.clone()).sinto(s)
+    }
+}
+
```

Comment:
I think this is OK for this PR, but then we should simplify those things, the predicate indirection is useless now (I'm not sure wether this was useful at any point now :sweat_smile:) and requires wrapping with dummy binders, which indeed is quite dangerous.

Reply:
To be clear: this PR does not introduce any new `Binder::dummy`, these were moved from somewhere else. I agree we should take the time to remove them at some point.

Reply:
Well, I meant the "PolyClauseKind" actually, not the `Predicate` (which is a bit a "PolyPredicateKind" ^^)

Reply:
I'm not sure what you mean, but we should definitely be more careful with binders

Reply:
Yes, let's see that later anyway